src/iceberg/type.cc

```diff
@@ -0,0 +1,314 @@
+TypeId MapType::type_id() const { return TypeId::kMap; }
+std::string MapType::ToString() const {
+  // XXX: work around Clang/libc++: "<{}>" in a format string appears to get
+  // parsed as {<>} or something; split up the format string to avoid that
+  std::string repr = "map<";
+
+  std::format_to(std::back_inserter(repr), "{}: {}", key(), value());
+  repr += ">";
+  return repr;
+}
+std::span<const SchemaField> MapType::fields() const { return fields_; }
+std::optional<std::reference_wrapper<const SchemaField>> MapType::GetFieldById(
+    int32_t field_id) const {
+  if (field_id == key().field_id()) {
+    return key();
+  } else if (field_id == value().field_id()) {
+    return value();
+  }
+  return std::nullopt;
+}
+std::optional<std::reference_wrapper<const SchemaField>> MapType::GetFieldByIndex(
+    int32_t index) const {
+  if (index == 0) {
+    return key();
+  } else if (index == 1) {
+    return value();
+  }
+  return std::nullopt;
+}
+std::optional<std::reference_wrapper<const SchemaField>> MapType::GetFieldByName(
+    std::string_view name) const {
+  if (name == kKeyName) {
+    return key();
+  } else if (name == kValueName) {
+    return value();
+  }
+  return std::nullopt;
+}
+bool MapType::Equals(const Type& other) const {
+  if (other.type_id() != TypeId::kMap) {
+    return false;
+  }
+  const auto& map = static_cast<const MapType&>(other);
+  return fields_ == map.fields_;
+}
+
+StructType::StructType(std::vector<SchemaField> fields) : fields_(std::move(fields)) {
+  size_t index = 0;
+  for (const auto& field : fields_) {
+    auto [it, inserted] = field_id_to_index_.try_emplace(field.field_id(), index);
+    if (!inserted) {
+      throw std::runtime_error(
+          std::format("StructType: duplicate field ID {} (field indices {} and {})",
+                      field.field_id(), it->second, index));
+    }
+
+    index++;
+  }
+}
+
+TypeId StructType::type_id() const { return TypeId::kStruct; }
+std::string StructType::ToString() const {
+  std::string repr = "struct<\n";
+  for (const auto& field : fields_) {
+    std::format_to(std::back_inserter(repr), "  {}\n", field);
+  }
+  repr += ">";
+  return repr;
+}
+std::span<const SchemaField> StructType::fields() const { return fields_; }
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldById(
+    int32_t field_id) const {
+  auto it = field_id_to_index_.find(field_id);
+  if (it == field_id_to_index_.end()) return std::nullopt;
+  return fields_[it->second];
+}
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldByIndex(
+    int32_t index) const {
+  if (index < 0 || index >= static_cast<int>(fields_.size())) {
+    return std::nullopt;
+  }
+  return fields_[index];
+}
+std::optional<std::reference_wrapper<const SchemaField>> StructType::GetFieldByName(
+    std::string_view name) const {
+  // TODO: what is the right behavior if there are duplicate names? (Are
```

Comment:
I found that the java and the rust implement has a lazy initialization and locked hash map in get field by name, but I don't understand the reason...

Reply:
The reason for the laziness, or the reason for the map? (I could change this to be lazy but tbh that introduces quite a bit of complication...)

Reply:
I prefer just do like this until someone need optimize this!